### PR TITLE
TypeSerializer for reliable passing types from collectors

### PR DESCRIPTION
**Pull request by MartinMystikJonas**
All type serialization/deserialization between object and array (passed from colelctors) is done by new service TypeSerializer

### CODE REVIEW

File: src/Type/TypeSerializer.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+     * @param TypeStringResolver $typeStringResolver
+     */
+    public function __construct(TypeStringResolver $typeStringResolver)
+    {
+        $this->typeStringResolver = $typeStringResolver;
+    }
+
+    /**
+     * @param Type $type
+     * @return array<string, string>
```
Comment (lulco) [inline src/Type/TypeSerializer.php:24]: shouldn't be `array{precise: string, basic: string, serialized?: string}`? And also everywhere where it is used in value object array types

File: src/Type/TypeSerializer.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+use PHPStan\Type\Type;
+use PHPStan\Type\VerbosityLevel;
+use Throwable;
+
+class TypeSerializer
+{
+    private TypeStringResolver $typeStringResolver;
+
+    /**
+     * @param TypeStringResolver $typeStringResolver
```
Comment (lulco) [inline src/Type/TypeSerializer.php]: this comment has no additional information

File: src/Type/TypeSerializer.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public function __construct(TypeStringResolver $typeStringResolver)
+    {
+        $this->typeStringResolver = $typeStringResolver;
+    }
+
+    /**
+     * @param Type $type
```
Comment (lulco) [inline src/Type/TypeSerializer.php]: this parameter comment has no additional information

File: src/Type/TypeSerializer.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            $typeAsArray['serialized'] = @serialize($type);
+        } catch (Throwable $e) {
+            // ignore
+        }
+        return $typeAsArray;
+    }
+
+    /**
+     * @param array<string, string> $typeAsArray
+     * @return Type
```
Comment (lulco) [inline src/Type/TypeSerializer.php]: no additional value

File: src/Type/TypeSerializer.php (hunk 1 above)
Comment (MartinMystikJonas) [reply src/Type/TypeSerializer.php:24]: My idea was taht exact format of serialized type is internal information of TypeSerializer - we might change/add keys and it never should be accessed outside this class

Comment (MartinMystikJonas) [reply src/Type/TypeSerializer.php:24]: I would either keep it this was or create type alias TypeAsArray that would be used across app? What would you preffer?

**Comment by MartinMystikJonas** [comment]
Should I merge it already or add that TypeAsArray type alias first ?

File: src/Type/TypeSerializer.php (hunk 1 above)
Comment (lulco) [reply src/Type/TypeSerializer.php:24]: Ok if we don’t need it outside of this class we can leave it string, string

**Review by lulco — APPROVED** [review]

**Comment by lulco** [comment]
> Should I merge it already or add that TypeAsArray type alias first ?

You can merge

**Pull request merged**